ideal_device/io/tlb_manager.h

```diff
@@ -0,0 +1,183 @@
+    virtual void release_tlb_window(TLBWindow& tlb_window) = 0;
+
+    // This is something that is probably going to need per arch implementation.
+    virtual TLBWindow get_dram_tlb_widow(uint32_t dram_channel, uint32_t addr = 0) = 0;
+
+    virtual TLBWindow release_dram_tlb_window(TLBWindow& tlb_window) = 0;
+protected:
+    // Are these constants supposed to be in per arch tlb manager class?
+    static constexpr uint32_t uc_window_start = 0;
+    static constexpr uint32_t uc_window_count = 0;
+
+    static constexpr uint32_t wc_window_start = 0;
+    static constexpr uint32_t wc_window_count = 0;
+
+    // dummy implementation
+    uint32_t map_core_to_tlb(tt_xy_pair core) {
+        return 0;
+    }
+}
+
+// This should represent how we do things today. We have whole BAR0/4 space
+// and just map the TLB we need.
+class StandardTLBManager : public TLBManager {
+
+public:
+    StandardTLBManager(PCIDevice* pci_device) : pci_device(pci_device) {}
+
+    TLBWindow get_tlb_window(tt_xy_pair core) override {
+        uint32_t tlb_index = map_core_to_tlb_index(core);
+        // This should be a WC window.
+        if (tlb_index < wc_window_start || tlb_index >= uc_window_start) {
+            // throw error.
+        }
+        wc_tlb_window_count[tlb_index - wc_window_start]++;
+
+        return get_tlb_window(tlb_index);
+    }
+    
+    // Get unused WC TLB window.
+    TLBWindow get_wc_tlb_window() override {
+        // Find unused UC TLB window.
+        for (uint32_t i = 0; i < wc_window_count; i++) {
+            if (wc_tlb_window_count[i] == 0) {
+                wc_tlb_window_count[i]++;
+                return get_tlb_window(wc_window_start + i);
+            }
+        }
+
+        // throw error.
+    }
+
+    // Get unused UC TLB window.
+    TLBWindow get_uc_tlb_window() override {
+        // Find unused UC TLB window.
+        for (uint32_t i = 0; i < uc_window_count; i++) {
+            if (uc_tlb_window_count[i] == 0) {
+                uc_tlb_window_count[i]++;
+                return get_tlb_window(uc_window_start + i);
+            }
+        }
+    }
+
+    void release_tlb_window(TLBWindow& tlb_window) override {
+        uint32_t tlb_index = tlb_window.index;
+        if (tlb_index >= wc_window_start && tlb_index < uc_window_start) {
+            // Do we need to reprogram it?
+            wc_tlb_window_count[tlb_index - wc_window_start]--;
+        } else if (tlb_index >= uc_window_start) {
+            // Do we need to reprogram it?
+            uc_tlb_window_count[tlb_index - uc_window_start]--;
+        } else {
+            // throw error.
+        }
+    }
+
+    TLBWindow get_dram_tlb_widow(uint32_t dram_channel, uint32_t addr = 0) override {
+        // We need to get DRAM core from DRAM channel.
+        return get_wc_tlb_window(tt_xy_pair(0, 0));
+    }
+
+    void release_dram_tlb_window(TLBWindow& tlb_window) override {
+        release_tlb_window(tlb_window);
+    }
+
+private:
+    PCIDevice* pci_device;
```

Comment:
Not sure if intentional, due to my renamings. But the reference should be to ttdevice, right?
We could have TTDevice->get_tlb_manager(managertype) , and then use it to do management. Not sure if then the user would get the ownership as well (gets unique ptr), and that resulting manager has pointer to trdevice so it can use its interface which covers arch specific code

Reply:
Probably it should be `TTDevice`, because we can get `PCIDevice` from `TTDevice` if needed.  There is a question on how do we manage TLBs. If we don't program TLBs in advance, the question is where are we going to program the TLBs. For programming the TLB we need bar0 and bar4 pointers, which are exposed through PCIDevice. `TLBManager` seems like a good place to do it. Generally, `TLBManager` logic could go to `TTDevice`, but `TTDevice` has `PCIDevice` in it, so to me it looks the same. `TLBManager` was just used to encapsulate the logic for managing TLBs.